external-crates/move/documentation/book/src/index-syntax.md

```diff
@@ -0,0 +1,297 @@
+let m_0_0 = mat[0, 0];
+```
+
+During compilation, the compiler will translate these into the appropriate function invocations
+based on the position and mutable usage of the expression:
+
+```move
+let mut mat = matrix::make_matrix(...);
+
+let m_0_0 = mat[0, 0];
+    // translates to copy matrix::borrow(&mat, 0, 0)
+
+let m_0_0 = &mat[0, 0];
+    // translates to matrix::borrow(&mat, 0, 0)
+
+let m_0_0 = &mut mat[0, 0];
+    // translates to matrix::borrow_mut(&mut mat, 0, 0)
+``
+
+Index expressions may also be intermixed with field accesses:
+
+```move
+public struct V { v: vector<u64> }
+
+public struct Vs { vs: vector<V> }
+
+fun borrow_first(input: &Vs): &u64 {
+    input.vs[0].v[0]
+    // translates to vector::borrow(vector::borrow(input.vs, 0).v, 0)
+}
+```
+
+### Index functions take flexible arguments
+
+Note that, aside from the definition and type limitations described in the rest of this chapter,
+there are no restrictions on the values that may be passed as index parameters, allowing for
+intricate programmatic behavior when using index syntax. For example, a data structure might wish to
+take a default value if the index is out of bounds:
+
+```
+#[syntax(index)]
+public fun borrow_or_set<Key: copy, Value: drop>(
+    input: &mut MTable<Key, Value>,
+    key: &Key,
+    default: Value
+): &mut Value {
+    if (contains(input, *key)) {
+        borrow(input, key)
+    } else {
+        insert(input, *key, default)
+        borrow(input, key)
+    }
+}
+```
+
+Indexing into `MTable` would then require the user provide a default value:
+
+```
+let string_key: String = ...;
+let mut table: MTable<String, u64> = m_table::make_table();
+let entry: &mut u64 = &mut table[string_key, 0];
+```
+
+This sort of extensible power allows developers to write precise index interfaces for their types,
+concretely enforcing bespoke behavior.
+
+
+## Defining Index Syntax Functions
+
+This powerful syntax form allows for all user-defined datatypes to behave in this way, assuming they
+adhere to the following rules:
+
+1. The `#[syntax(index)]` attribute is added to the designated functions defined in the same module
+   as the subject type.
+2. The designated functions have `public` visibility.
+3. The functions take a reference type as its "subject" type (its first argument) and returns a
+   matching references type (`mut` if the subject was `mut`).
+4. Each type has only a single mutable and single immutable definition.
+5. Immutable and mutable versions have type agreement:
+    - The subject types match, differing only in mutability.
+    - The return types match the mutability of their subject types.
+    - Type parameters, if present, have identical constraints between both versions.
+    - All parameters beyond the subject type are identical.
+
+These are described in greater detail, with additional examples, below.
+
+### Declaration
+
+To declare an index syntax method, you add the `#[syntax(index)]` attribute above the relevant
```

Comment:
Switching to second person here. I'd suggest using second person throughout (instead of "developers" or "users") but that can be addressed in a final pass. Adding this note so you can point out if you feel strongly one way or the other about the use of 2nd person.